persist/fs/digest.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"errors"
+	"hash"
+	"hash/adler32"
+	"io/ioutil"
+	"os"
+)
+
+var (
+	// errReadFewerThanExpectedBytes returned when number of bytes read is fewer than expected
+	errReadFewerThanExpectedBytes = errors.New("number of bytes read is fewer than expected")
+
+	// errCheckSumMismatch returned when the calculated checksum doesn't match the stored checksum
+	errCheckSumMismatch = errors.New("calculated checksum doesn't match stored checksum")
+)
+
+type fdWithDigest struct {
+	fd     *os.File
+	digest hash.Hash32
```

Comment:
There's a lot of non-encapsulated access to `digest` which makes it hard to read the methods without understanding the access patterns of users of this struct.  I'd be in preference of reducing direct access to fields on this struct so that one can understand the usage of the fields in isolation easier.